packages/contracts/test/fuzz/diamond/facets/StakingFacet.fuzz.t.sol

```diff
@@ -0,0 +1,150 @@
+    MockERC20 stakeToken;
+
+    address user = makeAddr("user");
+    address user2 = makeAddr("user2");
+
+    function setUp() public override {
+        super.setUp();
+
+        vm.prank(owner);
+        dollarManager = new UbiquityAlgorithmicDollarManager(owner);
+
+        vm.prank(owner);
+        rewardToken = new UbiquityGovernance(address(dollarManager));
+
+        stakeToken = new MockERC20("STK", "STK", 18);
+
+        // staking setup
+        vm.startPrank(admin);
+        stakingFacet.setGovernancePerBlock(1 ether);
+        stakingFacet.setGovernanceTreasuryDivider(5);
+        stakingFacet.setStakingRewardToken(address(rewardToken));
+        stakingFacet.setStakingStartBlock(block.number);
+        vm.stopPrank();
+
+        // owner grants diamond the "UBQ_MINTER_ROLE"
+        // NOTICE: in production environment the diamond contract already has the "UBQ_MINTER_ROLE" role
+        vm.prank(owner);
+        dollarManager.grantRole(keccak256("UBQ_MINTER_ROLE"), address(diamond));
+
+        // admin creates a new staking pool
+        vm.prank(admin);
+        stakingFacet.createStakingPool(
+            100, // allocation points
+            stakeToken,
+            true // whether to update all pools
+        );
+
+        // user approves diamond to spend STK tokens
+        vm.prank(user);
+        stakeToken.approve(address(diamond), type(uint256).max);
+
+        // user2 approves diamond to spend STK tokens
+        vm.prank(user2);
+        stakeToken.approve(address(diamond), type(uint256).max);
+    }
+
+    function testStaking_ShouldStakeAndUnstakeTokens(
+        uint256 stakeAmount,
+        uint256 blocksPassed,
+        uint256 governancePerBlock
+    ) public {
+        stakeAmount = bound(stakeAmount, 1, 100_000_000 ether);
+        blocksPassed = bound(blocksPassed, 1, 2628000 * 10); // max 10 years
+        governancePerBlock = bound(governancePerBlock, 1, 100_000_000 ether);
+
+        // admin sets governance per block
+        vm.prank(admin);
+        stakingFacet.setGovernancePerBlock(governancePerBlock);
+
+        // mint X STK tokens to users
+        stakeToken.mint(user, stakeAmount);
+        stakeToken.mint(user2, stakeAmount);
+
+        // before staking
+        LibStaking.UserInfo memory userInfo = stakingFacet.getStakingUserInfo(
+            0,
+            user
+        );
+        LibStaking.UserInfo memory userInfo2 = stakingFacet.getStakingUserInfo(
+            0,
+            user2
+        );
+        LibStaking.PoolInfo memory poolInfo = stakingFacet.getStakingPoolInfo(
+            0
+        );
+        assertEq(rewardToken.balanceOf(user), 0);
+        assertEq(rewardToken.balanceOf(user2), 0);
+        assertEq(stakeToken.balanceOf(address(stakingFacet)), 0);
+        assertEq(userInfo.amount, 0);
+        assertEq(userInfo2.amount, 0);
+        assertEq(poolInfo.amount, 0);
+
+        // users stake STK
+        vm.prank(user);
+        stakingFacet.stake(0, stakeAmount);
```

Comment:
would be nice to fuzz pool creation and stake on another fuzz test

Reply:
Added more fuzzing tests in https://github.com/ubiquity/ubiquity-dollar/pull/987/commits/9c7fef6dab4d04cebffabd43593e978a7a5816ce